vcpp-client-reference/CMakeLists.txt

```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required(VERSION 3.21.0)
+
+set(CMAKE_TOOLCHAIN_FILE
+    "${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake"
+    CACHE STRING "Vcpkg toolchain file")
+
+project(volition)
+
+find_package(gRPC CONFIG REQUIRED)
+find_package(Protobuf REQUIRED)
+
+add_library(optimizer-proto optimizer.proto)
+# Include the generated header files
+target_include_directories(optimizer-proto PUBLIC "${CMAKE_CURRENT_BINARY_DIR}")
+target_link_libraries(optimizer-proto PUBLIC gRPC::grpc++)
+# Generate Protobuf files
+protobuf_generate(TARGET optimizer-proto)
+# Generate gRPC files
+protobuf_generate(
+    TARGET optimizer-proto
+    LANGUAGE grpc
+    PLUGIN protoc-gen-grpc=$<TARGET_FILE:gRPC::grpc_cpp_plugin>
+    GENERATE_EXTENSIONS .grpc.pb.h .grpc.pb.cc)
+
+add_executable(vcpp-client-reference Source.cpp)
+target_link_libraries(vcpp-client-reference PRIVATE optimizer-proto)
```

Comment:
as per our talk and pesuant to my ability to learn cmake:

- `add_library` declares a kind of group that we'll attach input source and output binaries to
- the `set(CMAKE_COOLCHAIN_FILE ...)` is where we append the vcpkg magic that will put things like protobuf on cmakes internal PATH, given that we configured vpckg to pull protobuf/grpc/etc
- we need to tell the optimizer-proto library blob that its headers must include the grpc headers and its binaries must include the ... current binrary dir? cant we be more spcific here?

Reply:
- Yes, `add_library` creates a target.
- Right. `set(CMAKE_TOOLCHAIN_FILE ...)` is where the Vcpkg magic happens.
- `target_include_directories` attaches include directories to this target. The Protobuf and gRPC header files generated from the `.proto` file are created in `${CMAKE_CURRENT_BINARY_DIR}`, so we add this include path to this target, or else the compiler won't be able to find the generated header files. Here the `PUBLIC` flag is used so these path settings are propagated to subsequent targets that link against it. In this sense, CMake targets carry and forward their own dependencies with them.